Readme (#212)
insightsengineering/tlg-catalog: package/README.md

# tlg.catalog.pkg

This is a dummy package used for testing purposes. For the actual content please navigate to the [book directory](../book).

> The package itself is relatively empty and most of the logic lives in the `tests/` subdirectory. The main goal is to reuse state of art GitHub Actions and other CI/CD tools designed for package development. This includes `lintr`, `styler`, `spelling`, and `testthat2` snapshot testing.

Reviewer: most of the logic lives in the `tests/` subdirectory - too colloquial

## Regression tests

### Overview

> The book component is symlinked into `tests/testthat/_book` subdirectory (no need to copy&paste). As a part of test setup, the whole book will be regenerated. A dedicated `knitr` hooks will be invoked to save relevant output objects as `.Rds` files. During actual testing, these files are loaded and compared with the expected ones using `testthat2` snapshot testing.

Author: Yes you are. But that's a good thing as you are a perfect reviewer :) I changed this a little - please have a look!

A _standard_ render will be unaffected. These hooks will be used only during testing.

> ### How to use?

Reviewer: How to use what?

* At the beggining of the article, add the following: `{{< include ../../test-utils/envir_hook.qmd >}}`. It will register the `knitr` hooks for further use.
* For code chunks without output - do nothing.
* For code chunks with output:

>   * Inside chunk, make sure you save the output into the object (e.g. `table <- foo(...); table` instead of `foo(...)`). Don't forget to print the object at the end of the chunk to make it visible in the article.

Reviewer: move example to end of paragraph

  * In the chunk header, use a dedicated `test` chunk template giving previously described object name a new, unique(!) name using following syntax: `test = list(<new name> = "<old name>")` (e.g. use `table` object from the current chunk and name it as `result_v1`: `{r, test = list(result_v1 = "table")}`).
* At the end of the article (after all chunks with outputs), add `{{< include ../../test-utils/save_results.qmd >}}` to save all objects as `.Rds` files.

> This logic had been implemented in many articles already. Please refer to the existing ones for examples.

Reviewer: ```suggestion
This logic has been implemented in many articles already. Please refer to the existing ones for examples.
```

"Many"? Meaning some but not all?

Author: Presumably all but there might be some exceptions. It's nowhere enforced.

If there are no outputs to be tested, you can skip the whole process. The whole article (all code chunks) will be rendered but no output will be saved and included in tests. In such case, for performance reasons, it's recommended to use a dedicated `skip_if_testing` chunk template, e.g. `{r, skip_if_testing = TRUE}`. This will prevent a given code chunk from being executed during testing.

> ### How it works?

Reviewer: Please use proper English.

This functionality relies on [`include` Quarto shortcode](https://quarto.org/docs/authoring/includes.html), [`knitr` chunk hooks](https://yihui.org/knitr/hooks/), [`knitr` option templates](https://yihui.org/knitr/options/#option-templates) and [`testthat2` snapshot testing](https://testthat.r-lib.org/articles/snapshotting.html).
When loading the hooks, a new environment type of variable `tenv` is created. Each time a `test` template is used, it will take the object from the current environment (using provided name) and save it in the `tenv` environment. At the end, all objects from `tenv` are saved as `.Rds` file.

During testing, a for loop over articles is run. If `.Rds` file is missing, the test is skipped. If it's present, it's loaded and its print output is compared with the expected one.

> 

Reviewer: This is not entirely clear to me, perhaps I am missing some context.

> ### How to update indivudal snapshot(s)?

Reviewer: ```suggestion
### How to update individual snapshots?
```

> Rendering the whole book might be time consuming. To update a single snapshot, you can temporarily change `test/testthat/setup.R` file to render only the article you need. A vector of articles is also possible. After the change, run `devtools::test()` to run tests and invoke dedicated hooks - they are working only during testing. After the test, review the snapshot changes (if any) and don't forget to revert the changes in `setup.R` file.

Reviewer: This could be phrased better. I understand that the articles to be rendered are the ones listed in `created_files` in `tests/testthat/setup.R`.

# tlg.catalog.pkg

This is a dummy package used for testing purposes. For the actual content please navigate to the [book directory](../book).

The majority of the logic resides in the `tests/` subdirectory, while the package itself remains minimal.

> The main purpose of it is to reuse state of art GitHub Actions and other CI/CD tools for package development on the book content.

Reviewer: ```suggestion
The main purpose of it is to reuse state of the art GitHub Actions and other CI/CD tools for package development on the book content.
```

This also includes `testthat2` snapshot testing for regression tests.

## Regression tests

### Overview

During tests, as a part of [test setup](./tests/testthat/setup.R), the whole book will be re-rendered.
This is possible because the content of the book is symlinked to the tests directory (symbolic link of `../book` as a `tests/testthat/_book`).

> There are dedicated `knitr` hooks and templates available and if used (see below section of how to use them), they will save relevant output objects as `.Rds` files.

Reviewer: ```suggestion
There are dedicated `knitr` hooks and templates available that will save relevant output objects as `.Rds` files when used (see below).
```

> During actual testing, these files are loaded and compared with the expected ones using `testthat2` snapshot testing.

Reviewer: ```suggestion
When tests are run, these files are loaded and compared with the expected ones using `testthat2` snapshot testing.
```

A _standard_ render will be unaffected. These hooks will be used only during testing.

### How to use `knitr` hooks and templates?

As an article author, you need to (i) register the hooks, (ii) use them where appropriate and (iii) save the results.

* At the beggining of the article, add the following: `{{< include ../../test-utils/envir_hook.qmd >}}`. It will register the `knitr` hooks for further use.
* For code chunks without output - do nothing.

> * For code chunks with output:

Reviewer: It would be nice to break lines before examples it that's possible.

Author: Thanks. I will check this one. Here, I need to comply with markdownlint which is tricky for lists

Reviewer: I know :\ It's not crucial.

Author: ok done

>   * Inside chunk, make sure you save the output into the object and don't forget to print the object at the end of the chunk to make it visible in the article. Example: `table <- foo(...); table` instead of `foo(...)`.

Reviewer: ```suggestion
  * Inside chunk, make sure you assign the output to a variable and print the variable at the end of the chunk to make the output visible in the article. Example: `table <- foo(...); table` instead of `foo(...)`.
```

>   * In the chunk header, use a dedicated `test` chunk template giving previously described object name a new, unique(!) name using following syntax: `test = list(<new name> = "<old name>")`. Example: `{r, test = list(result_v1 = "table")}` to use `table` and name it `result_v1`.

Reviewer: ```suggestion
  * In the chunk header, provide a `test` chunk template listing the aforementioned variable with a unique name, using following syntax: `test = list(<name> = "<variable name>")`. Example: `{r, test = list(result_v1 = "table")}` to use `table` and name it `result_v1`.
```

Author: Here I think that "use" will be better than "provide". Please accept my edit of your message and they will apply it.

* At the end of the article (after all chunks with outputs), add `{{< include ../../test-utils/save_results.qmd >}}` to save all objects as `.Rds` files.

This logic has been implemented in many articles already. Please refer to the existing ones for examples.

If there are no outputs to be tested, you can skip the whole process.
The whole article (all code chunks) will be rendered but no output will be saved and included in tests.

> In such case, for performance reasons, it's recommended to use a dedicated `skip_if_testing` chunk template, e.g. `{r, skip_if_testing = TRUE}`.

Reviewer: ```suggestion
In such a case, for performance reasons, it is recommended to use a dedicated `skip_if_testing` chunk template, _e.g._ `{r, skip_if_testing = TRUE}`.
```

This will exclude a given code chunk from being executed during testing.

> ### How does it work

Reviewer: ```suggestion
### How does it work?
```

This functionality relies on [`include` Quarto shortcode](https://quarto.org/docs/authoring/includes.html), [`knitr` chunk hooks](https://yihui.org/knitr/hooks/), [`knitr` option templates](https://yihui.org/knitr/options/#option-templates) and [`testthat2` snapshot testing](https://testthat.r-lib.org/articles/snapshotting.html).

> After including `envir_hook.qmd` file, there are a few new `knitr` hooks registered.

Reviewer: I assume the inclusion of this file is explained in the links above but that is not readily apparent.

Author: Above links are for tools & techniques (generally speaking). We _implementated_ (or directly _used_) them. You won't find there `envir_hook` or any other custom names we have. On the other hand, you will find the manual for `{{ include ... }}` shortcode, `knitr::knit_hooks`, `knitr::opts_template` and others.

If you are referring to the _inclusion_ then it's the first link -> https://quarto.org/docs/authoring/includes.html.

Does that answer your question?

Author: Few lines above we have:

> * At the beggining of the article, add the following: `{{< include ../../test-utils/envir_hook.qmd >}}`. It will register the `knitr` hooks for further use.

Should I repeat it / elaborate more on this section as well?

Reviewer: Ooooh, I missed that! Honestly, I didn't read the line that is meant to be pasted into the article.

How about this?

```
Registering `knitr` hooks alone will not produce any visible changes in the article.
However, when evaluating code chunks where a test template is used, the object specified in the template will be copied to a newly created environment, `tenv`.
When evaluation reaches the end of the article, `tenv` is saved in an .Rds file.
```

Author: We are going in the good direction. Save also needs to be explicitly run - you don't have to know this.

How about this:
- move `It will register the knitr hooks for further use.` sentence from "How to use" into "How does it work" (this paragraph) as it belongs more there
- have the following:
```
Including `envir_hook.qmd` beforehand will register a few `knitr` hooks for the use throughout the article.
This will not produce any visible changes to the content.
However, when evaluating code chunks where a `test` template is used, the object specified will be copied to a newly created, internal environment.
When `save_results.qmd` is included, the environment object is saved in an `.Rds` file.
```

~When I look at this - I guess we can also remove `tenv` as this is not critical to understand. It's more a detail of implementation.~
OK removed

Reviewer: There is still a disconnect between "including envir_hook.qmd" and the line inserted into the file. Again, if I just have to copy it there, I will likely not read it.

How about tying actions to the points enumerated above?
`As an article author, you need to (i) register the hooks, (ii) use them where appropriate and (iii) save the results.`

Something like this:

```
(i) register hooks

- At the beggining of the article, add the following: {{< include ../../test-utils/envir_hook.qmd >}}. It will register the knitr hooks for further use.

(ii) use them where appropriate

- For code chunks with output:

+ Inside chunk, make sure you assign the output to a variable and print the variable at the end of the chunk to make the output visible in the article.

Example: table <- foo(...); table instead of foo(...).

+ In the chunk header, use the test chunk template listing the aforementioned variable with a unique name, using following syntax: test = list(<name> = "<variable name>").

Example: {r, test = list(result_v1 = "table")} to use table and name it result_v1.

- For code chunks without output - do nothing.

(iii) save the results

At the end of the article (after all chunks with outputs), add {{< include ../../test-utils/save_results.qmd >}} to save all objects as .Rds files.
```

Author: I have pushed a new version. Would you mind to have a look? 

There are two unresolved conversations left. This as well as https://github.com/insightsengineering/tlg-catalog/pull/212#discussion_r1495448398

Reviewer: Excellent. One last detail would be 
```
It will register the `knitr` hooks for further use and would not produce any visible changes to the content.
```
"Would not" bothers me a little. "Will not"?

Author: good catch - pushed the update

This will not produce any visible changes in the article.
Each time a `test` chunk template is used, it will take the specified object from the current environment and save it in the newly created environment type of object `tenv`.

> At the end, `tenv` object is saved as `.Rds` file.

Reviewer: ```suggestion
At the end, the `tenv` object is saved in an `.Rds` file.
```

> During testing, a for loop over articles is run.

Reviewer: ```suggestion
During testing, a `for` loop is run over all articles.
```

> If `.Rds` file is missing, the test is skipped.

Reviewer: Is this right?
```suggestion
If the `.Rds` file for an article is missing, the test is skipped.
```

> If it's present, it's loaded and its print output is compared with the expected one.

Reviewer: No contractions, please.
```suggestion
If it is present, it is loaded and its print output is compared with the expected one.
```

Which output is the expected one, actually? The one in the file or the nascent one?

Author: There is a check against newly created output (temporarily save to the file) against the one that had been previously committed into some subdirectory inside `tests/testthat/...`

### How to update individual snapshots?

Rendering the whole book might be time consuming.
To update a single snapshot, you can temporarily change `test/testthat/setup.R` file to render only the article you need.
A vector of articles is also possible.

> After the change, run `devtools::test()` to run tests and invoke dedicated hooks - they are working only during testing.

Reviewer: ```suggestion
After the change, run `devtools::test()` to run tests and invoke dedicated hooks - they work only during testing.
```

> After the test, review the snapshot changes (if any) and don't forget to revert the changes in `setup.R` file.

Reviewer: ```suggestion
After the test, review snapshot deviations (if any) and remember to revert the changes in `setup.R`.
```

> Sometimes rendering the whole book is unavoidable. For this, run `devtools::test()` without any additional changes.

Reviewer: ```suggestion
Sometimes rendering the whole book cannot be avoided. In those cases, run `devtools::test()` without any additional changes.
```
